// component-injector/__ivy_ngcc__/fesm2015/component-injector.js
import { Injectable, ComponentFactoryResolver, ɵɵdefineInjectable, ɵɵinject, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
class ComponentInjector {
    /**
     * @param {?} resolver
     */
    constructor(resolver) {
        this.resolver = resolver;
    }
    /**
     * @param {?} componentsList
     * @return {?}
     */
    setComponentFactories(componentsList) {
        if (componentsList) {
            componentsList.forEach((/**
             * @param {?} component
             * @return {?}
             */
            (component) => {
                /** @type {?} */
                let factory = this.resolver.resolveComponentFactory(component);
                if (factory) {
                    ComponentInjector.entryComponentsFactories.push(factory);
                }
            }));
        }
    }
    /**
     * @param {?} container
     * @param {?} componentSelector
     * @return {?}
     */
    inject(container, componentSelector) {
        /** @type {?} */
        let componentFactory = this.getComponentFactory(componentSelector);
        if (componentFactory) {
            return this.injectComponentFactory(container, componentFactory);
        }
        else {
            console.warn('Component [' + componentSelector + '] cannot be found! ' +
                'Make sure it is included in the `entryComponents` list.');
        }
        return null;
    }
    /**
     * @param {?} componentRef
     * @param {?} properties
     * @return {?}
     */
    setProperties(componentRef, properties) {
        if (componentRef) {
            for (let property in properties) {
                if (properties.hasOwnProperty(property)) {
                    componentRef.instance[property] = properties[property];
                }
            }
        }
    }
    /**
     * @param {?} componentRef
     * @return {?}
     */
    remove(componentRef) {
        if (componentRef) {
            componentRef.destroy();
        }
        componentRef = null;
    }
    /**
     * @protected
     * @param {?} componentSelector
     * @return {?}
     */
    getComponentFactory(componentSelector) {
        /** @type {?} */
        let factories = ComponentInjector.entryComponentsFactories;
        /** @type {?} */
        let componentFactory = factories.find((/**
         * @param {?} component
         * @return {?}
         */
        (component) => {
            return component.selector === componentSelector;
        }));
        return componentFactory ? componentFactory : null;
    }
    /**
     * @protected
     * @param {?} container
     * @param {?} componentFactory
     * @return {?}
     */
    injectComponentFactory(container, componentFactory) {
        return container.createComponent(componentFactory);
    }
}
ComponentInjector.ɵfac = function ComponentInjector_Factory(t) { return new (t || ComponentInjector)(ɵngcc0.ɵɵinject(ɵngcc0.ComponentFactoryResolver)); };
ComponentInjector.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ComponentInjector, factory: ComponentInjector.ɵfac, providedIn: 'root' });
ComponentInjector.entryComponentsFactories = [];
/** @nocollapse */
ComponentInjector.ctorParameters = () => [
    { type: ComponentFactoryResolver }
];
/** @nocollapse */ ComponentInjector.ngInjectableDef = ɵɵdefineInjectable({ factory: function ComponentInjector_Factory() { return new ComponentInjector(ɵɵinject(ComponentFactoryResolver)); }, token: ComponentInjector, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ComponentInjector, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc0.ComponentFactoryResolver }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ComponentInjectorModule {
}
ComponentInjectorModule.ɵfac = function ComponentInjectorModule_Factory(t) { return new (t || ComponentInjectorModule)(); };
ComponentInjectorModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ComponentInjectorModule });
ComponentInjectorModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ComponentInjectorModule, [{
        type: NgModule,
        args: [{
                declarations: [],
                imports: [],
                exports: []
            }]
    }], null, null); })();

export { ComponentInjector, ComponentInjectorModule };
