// ng2-modal-module/__ivy_ngcc__/fesm2015/ng2-modal-module.js
import { PubSubDistinct } from 'pubsub-distinct';
import { Component, ViewChild, ViewContainerRef, Input, NgModule } from '@angular/core';
import { ComponentInjector } from 'component-injector';
import { XDomUtil } from 'xdom-util';
import { CommonModule } from '@angular/common';
import { TrustHtmlModule } from 'trust-html';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'component-injector';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'trust-html';

const _c0 = ["modalWindow"];
const _c1 = ["injectContainer"];
function Ng2ModalWindowComponent_div_15_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 16);
    ɵngcc0.ɵɵlistener("click", function Ng2ModalWindowComponent_div_15_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.cancelAction(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1("", ctx_r3.properties.buttons.cancel.label, " ");
} }
function Ng2ModalWindowComponent_div_15_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 17);
    ɵngcc0.ɵɵlistener("click", function Ng2ModalWindowComponent_div_15_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.successAction(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1("", ctx_r4.properties.buttons.success.label, " ");
} }
function Ng2ModalWindowComponent_div_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵtemplate(1, Ng2ModalWindowComponent_div_15_button_1_Template, 2, 1, "button", 14);
    ɵngcc0.ɵɵtemplate(2, Ng2ModalWindowComponent_div_15_button_2_Template, 2, 1, "button", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.properties.buttons.cancel == null ? null : ctx_r2.properties.buttons.cancel.visible);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.properties.buttons.success == null ? null : ctx_r2.properties.buttons.success.visible);
} }
class Ng2ModalWindow {
    /**
     * @param {?} modalId
     * @param {?=} options
     * @return {?}
     */
    static showModal(modalId, options = {}) {
        options.show = true;
        options.hide = null;
        PubSubDistinct.publishDistinct(modalId, options);
    }
    /**
     * @param {?} modalId
     * @return {?}
     */
    static hideModal(modalId) {
        /** @type {?} */
        let options = { hide: true };
        PubSubDistinct.publishDistinct(modalId, options);
    }
    /**
     * @param {?} eventsList
     * @return {?}
     */
    static resetEventsSubscribers(eventsList) {
        eventsList.forEach((/**
         * @param {?} eventName
         * @return {?}
         */
        (eventName) => {
            if (eventName && PubSubDistinct.hasSubscribers(eventName)) {
                PubSubDistinct.dispose(eventName);
            }
        }));
    }
    /**
     * @param {?} eventName
     * @param {?} callback
     * @return {?}
     */
    static subscribe(eventName, callback) {
        PubSubDistinct.subscribe(eventName, callback);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class Ng2ModalWindowComponent {
    /**
     * @param {?} componentInjector
     */
    constructor(componentInjector) {
        this.componentInjector = componentInjector;
        this.animationClass = 'fade';
        this.showModalClass = 'in';
        this.bodyOpenModalClass = 'modal-open';
        this.defaultProperties = {
            title: '',
            show: false,
            showEvent: false,
            hide: false,
            componentSelector: false,
            componentInputs: false,
            htmlContent: '',
            overlayClick: true,
            customClass: '',
            buttons: {
                visible: true,
                cancel: {
                    visible: true,
                    label: 'Cancel',
                    event: false
                },
                success: {
                    visible: true,
                    label: 'Save',
                    event: false
                }
            }
        };
        this.properties = {};
    }
    /**
     * @param {?} eventName
     * @return {?}
     */
    set id(eventName) {
        if (eventName) {
            this.eventName = eventName;
            // remove previous subscription and create new one
            this.unsubscribe();
            this.subscribeToEvent();
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.unsubscribe();
        this.resetInjectedComponent();
    }
    /**
     * @return {?}
     */
    cancelAction() {
        if (this.properties.buttons.cancel.event) {
            PubSubDistinct.publishDistinct(this.properties.buttons.cancel.event, true);
        }
        this.hide();
    }
    /**
     * @return {?}
     */
    successAction() {
        if (this.properties.buttons.success.event) {
            PubSubDistinct.publishDistinct(this.properties.buttons.success.event, true);
        }
        else {
            this.hide();
        }
    }
    /**
     * @return {?}
     */
    overlayClick() {
        if (this.properties.overlayClick) {
            this.cancelAction();
        }
    }
    /**
     * @protected
     * @return {?}
     */
    show() {
        // add class to modal DOM element to make it visible
        /** @type {?} */
        let modalDom = this.modalWindow.nativeElement;
        if (!XDomUtil.hasClass(modalDom, this.showModalClass)) {
            XDomUtil.addClass(modalDom, this.showModalClass);
        }
        // add class to modal body to disable the scrolling
        /** @type {?} */
        let body = document.querySelector('body');
        if (!XDomUtil.hasClass(body, this.bodyOpenModalClass)) {
            XDomUtil.addClass(body, this.bodyOpenModalClass);
        }
        if (this.properties.showEvent) {
            PubSubDistinct.publishDistinct(this.properties.showEvent, true);
        }
    }
    /**
     * @protected
     * @return {?}
     */
    hide() {
        /** @type {?} */
        let element = this.modalWindow.nativeElement;
        XDomUtil.removeClass(element, this.showModalClass);
        /** @type {?} */
        let body = document.querySelector('body');
        XDomUtil.removeClass(body, this.bodyOpenModalClass);
    }
    /**
     * @protected
     * @return {?}
     */
    subscribeToEvent() {
        this.eventSubscriber = PubSubDistinct.subscribe(this.eventName, (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            this.initModal(data);
        }));
    }
    /**
     * @protected
     * @param {?} properties
     * @return {?}
     */
    initModal(properties) {
        if (properties.show) {
            // remove previously injected component
            this.resetInjectedComponent();
            // reset the properties
            this.setProperties(properties);
            // inject component
            if (this.properties.componentSelector) {
                this.injectedComponentRef = this.injectComponent(this.properties.componentSelector);
                // set the components properties
                this.setComponentProperties();
            }
            // display the modal
            this.show();
            // reset modal event subscriber
            this.resetModalEventSubscriber();
        }
        else if (properties.hide) {
            // remove previously injected component
            this.resetInjectedComponent();
            // reset the properties
            this.setProperties(properties);
            // hide the modal
            this.hide();
            // reset modal event subscriber
            this.resetModalEventSubscriber();
        }
    }
    /**
     * @protected
     * @return {?}
     */
    unsubscribe() {
        if (this.eventSubscriber) {
            PubSubDistinct.unsubscribe(this.eventSubscriber);
        }
    }
    /**
     * @protected
     * @param {?} componentSelector
     * @return {?}
     */
    injectComponent(componentSelector) {
        /** @type {?} */
        let result;
        if (componentSelector) {
            result = this.componentInjector.inject(this.injectContainer, componentSelector);
        }
        return result;
    }
    /**
     * @protected
     * @return {?}
     */
    setComponentProperties() {
        if (this.properties.componentInputs && this.injectedComponentRef) {
            this.componentInjector.setProperties(this.injectedComponentRef, this.properties.componentInputs);
        }
    }
    /**
     * @private
     * @return {?}
     */
    resetInjectedComponent() {
        if (this.injectedComponentRef) {
            this.componentInjector.remove(this.injectedComponentRef);
            this.injectedComponentRef = null;
        }
    }
    /**
     * @private
     * @param {?} properties
     * @return {?}
     */
    setProperties(properties) {
        this.properties = Object.assign({}, this.defaultProperties, properties);
    }
    /**
     * @private
     * @return {?}
     */
    resetModalEventSubscriber() {
        // reset modal show/hide display
        PubSubDistinct.publishDistinct(this.eventName, {});
    }
}
Ng2ModalWindowComponent.ɵfac = function Ng2ModalWindowComponent_Factory(t) { return new (t || Ng2ModalWindowComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ComponentInjector)); };
Ng2ModalWindowComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: Ng2ModalWindowComponent, selectors: [["ng2-modal-window"]], viewQuery: function Ng2ModalWindowComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
        ɵngcc0.ɵɵviewQuery(_c1, 7, ViewContainerRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalWindow = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.injectContainer = _t.first);
    } }, inputs: { animationClass: "animationClass", id: "id" }, decls: 16, vars: 9, consts: [["tabindex", "-1", "role", "dialog", "id", "myModal", 1, "ko-modal-window", "modal", 3, "ngClass"], ["modalWindow", ""], [1, "ko-modal-overlay", 3, "click"], ["role", "document"], [1, "modal-content"], [1, "modal-header"], ["type", "button", "aria-label", "Close", 1, "close", 3, "click"], ["aria-hidden", "true"], [1, "modal-title"], [1, "modal-body"], [3, "innerHTML"], ["injectContainer", ""], ["class", "modal-footer", 4, "ngIf"], [1, "modal-footer"], ["type", "button", "class", "btn btn-default", 3, "click", 4, "ngIf"], ["type", "button", "class", "btn btn-primary", 3, "click", 4, "ngIf"], ["type", "button", 1, "btn", "btn-default", 3, "click"], ["type", "button", 1, "btn", "btn-primary", 3, "click"]], template: function Ng2ModalWindowComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵlistener("click", function Ng2ModalWindowComponent_Template_div_click_2_listener() { return ctx.overlayClick(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵelementStart(5, "div", 5);
        ɵngcc0.ɵɵelementStart(6, "button", 6);
        ɵngcc0.ɵɵlistener("click", function Ng2ModalWindowComponent_Template_button_click_6_listener() { return ctx.cancelAction(); });
        ɵngcc0.ɵɵelementStart(7, "span", 7);
        ɵngcc0.ɵɵtext(8, "\u00D7");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "h4", 8);
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 9);
        ɵngcc0.ɵɵelement(12, "div", 10, 11);
        ɵngcc0.ɵɵpipe(14, "trustHtml");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(15, Ng2ModalWindowComponent_div_15_Template, 3, 2, "div", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ctx.animationClass);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵclassMapInterpolate1("modal-dialog ", ctx.properties.customClass, "");
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ctx.properties.title);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind1(14, 7, ctx.properties.htmlContent), ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.properties == null ? null : ctx.properties.buttons == null ? null : ctx.properties.buttons.visible);
    } }, directives: [ɵngcc2.NgClass, ɵngcc2.NgIf], pipes: [ɵngcc3.TrustHtmlPipe], styles: [".ko-modal-window[_ngcontent-%COMP%]{background-color:rgba(0,0,0,.5)}.ko-modal-window.in[_ngcontent-%COMP%]{display:block}.ko-modal-open[_ngcontent-%COMP%]{overflow:hidden}.ko-modal-overlay[_ngcontent-%COMP%]{position:fixed;width:100%;height:100%;left:0;top:0}"] });
/** @nocollapse */
Ng2ModalWindowComponent.ctorParameters = () => [
    { type: ComponentInjector }
];
Ng2ModalWindowComponent.propDecorators = {
    modalWindow: [{ type: ViewChild, args: ['modalWindow', { static: true },] }],
    injectContainer: [{ type: ViewChild, args: ['injectContainer', { read: ViewContainerRef, static: true },] }],
    animationClass: [{ type: Input }],
    id: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng2ModalWindowComponent, [{
        type: Component,
        args: [{
                selector: 'ng2-modal-window',
                template: "<div class=\"ko-modal-window modal\"  [ngClass]=\"animationClass\" tabindex=\"-1\" role=\"dialog\" id=\"myModal\" #modalWindow>\n  <div class=\"ko-modal-overlay\" (click)=\"overlayClick()\"></div>\n  <div class=\"modal-dialog {{properties.customClass}}\" role=\"document\">\n    <div class=\"modal-content\">\n      <div class=\"modal-header\">\n        <button type=\"button\" class=\"close\" aria-label=\"Close\" (click)=\"cancelAction()\">\n          <span aria-hidden=\"true\">&times;</span>\n        </button>\n        <h4 class=\"modal-title\">{{properties.title}}</h4>\n      </div>\n      <div class=\"modal-body\">\n        <div #injectContainer [innerHTML]=\"properties.htmlContent | trustHtml\"></div>\n      </div>\n      <div class=\"modal-footer\" *ngIf=\"properties?.buttons?.visible\">\n        <button type=\"button\" class=\"btn btn-default\" (click)=\"cancelAction()\"\n                *ngIf=\"properties.buttons.cancel?.visible\">{{properties.buttons.cancel.label}}\n        </button>\n        <button type=\"button\" class=\"btn btn-primary\" (click)=\"successAction()\"\n                *ngIf=\"properties.buttons.success?.visible\">{{properties.buttons.success.label}}\n        </button>\n      </div>\n    </div><!-- /.modal-content -->\n  </div><!-- /.modal-dialog -->\n</div><!-- /.modal -->\n",
                styles: [".ko-modal-window{background-color:rgba(0,0,0,.5)}.ko-modal-window.in{display:block}.ko-modal-open{overflow:hidden}.ko-modal-overlay{position:fixed;width:100%;height:100%;left:0;top:0}"]
            }]
    }], function () { return [{ type: ɵngcc1.ComponentInjector }]; }, { animationClass: [{
            type: Input
        }], id: [{
            type: Input
        }], modalWindow: [{
            type: ViewChild,
            args: ['modalWindow', { static: true }]
        }], injectContainer: [{
            type: ViewChild,
            args: ['injectContainer', { read: ViewContainerRef, static: true }]
        }] }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ModalModule {
}
ModalModule.ɵfac = function ModalModule_Factory(t) { return new (t || ModalModule)(); };
ModalModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ModalModule });
ModalModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            TrustHtmlModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ModalModule, [{
        type: NgModule,
        args: [{
                declarations: [Ng2ModalWindowComponent],
                imports: [
                    CommonModule,
                    TrustHtmlModule
                ],
                exports: [Ng2ModalWindowComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ModalModule, { declarations: function () { return [Ng2ModalWindowComponent]; }, imports: function () { return [CommonModule,
        TrustHtmlModule]; }, exports: function () { return [Ng2ModalWindowComponent]; } }); })();

export { ModalModule, Ng2ModalWindow, Ng2ModalWindowComponent };
